File: initramfs.go
package lnx

import (
	"bytes"
	"fmt"
	"os"
	"path/filepath"
)

// InitBinary must be set by the embedding binary (via go:embed).
// The library itself does not embed the init binary — the caller provides it.
var InitBinary []byte

// WriteInitramfsTo creates a cpio-format initramfs and is exported for testing.
func WriteInitramfsTo(dir string) (string, error) {
	return writeInitramfs(dir)
}

func writeInitramfs(dir string) (string, error) {
	if len(InitBinary) == 0 {
		return "", fmt.Errorf("lnx.InitBinary not set; embed the guest init binary and assign it")
	}

	path := filepath.Join(dir, "initramfs.cpio")

	var buf bytes.Buffer
	writeCpioEntry(&buf, "init", InitBinary, 0100755)
	writeCpioEntry(&buf, "TRAILER!!!", nil, 0)
	if pad := buf.Len() % 512; pad != 0 {
		buf.Write(make([]byte, 512-pad))
	}

	if err := os.WriteFile(path, buf.Bytes(), 0644); err != nil {
		return "", err
	}
	return path, nil
}

// writeCpioEntry writes a single entry in cpio "newc" format.
func writeCpioEntry(buf *bytes.Buffer, name string, data []byte, mode uint32) {
	nameBytes := append([]byte(name), 0)
	hdr := fmt.Sprintf(
		"070701"+
			"%08X%08X%08X%08X%08X%08X%08X%08X%08X%08X%08X%08X%08X",
		1,              // inode
		mode,           // mode
		0,              // uid
		0,              // gid
		1,              // nlink
		0,              // mtime
		len(data),      // filesize
		0,              // devmajor
		0,              // devminor
		0,              // rdevmajor
		0,              // rdevminor
		len(nameBytes), // namesize
		0,              // check
	)
	buf.WriteString(hdr)
	buf.Write(nameBytes)
	if hdrLen := len(hdr) + len(nameBytes); hdrLen%4 != 0 {
		buf.Write(make([]byte, 4-hdrLen%4))
	}
	if data != nil {
		buf.Write(data)
		if len(data)%4 != 0 {
			buf.Write(make([]byte, 4-len(data)%4))
		}
	}
}

File: src/initramfs.rs
use std::{
    fs,
    io::Write,
    path::{Path, PathBuf},
};

use anyhow::{Context, Result};

pub fn write_from_agent(
    agent: &[u8],
    agent_source_stamp: &str,
    dir: PathBuf,
) -> Result<(PathBuf, bool)> {
    fs::create_dir_all(&dir).with_context(|| format!("create {}", dir.display()))?;
    let path = dir.join("initramfs.cpio");
    let stamp_path = dir.join("initramfs.stamp");
    let stamp = stamp(agent_source_stamp);

    if path.exists()
        && stamp_path.exists()
        && fs::read_to_string(&stamp_path).ok().as_deref() == Some(stamp.as_str())
    {
        return Ok((path, false));
    }

    write_agent(agent, &path)?;
    fs::write(&stamp_path, stamp).with_context(|| format!("write {}", stamp_path.display()))?;
    Ok((path, true))
}

fn write_agent(agent: &[u8], path: &Path) -> Result<()> {
    let mut buf = Vec::new();
    entry(&mut buf, "init", agent, 0o100755)?;
    entry(&mut buf, "lnx-agent", agent, 0o100755)?;
    entry(&mut buf, "lnxctl", agent, 0o100755)?;
    entry(&mut buf, "TRAILER!!!", &[], 0)?;
    if buf.len() % 512 != 0 {
        buf.resize(buf.len() + (512 - buf.len() % 512), 0);
    }
    fs::write(path, buf).with_context(|| format!("write {}", path.display()))?;
    Ok(())
}

fn stamp(agent_source_stamp: &str) -> String {
    format!("source={agent_source_stamp}\n")
}

fn entry(buf: &mut Vec<u8>, name: &str, data: &[u8], mode: u32) -> Result<()> {
    let namesize = name.len() + 1;
    write!(
        buf,
        "070701\
         {ino:08X}{mode:08X}{uid:08X}{gid:08X}{nlink:08X}{mtime:08X}{filesize:08X}\
         {devmajor:08X}{devminor:08X}{rdevmajor:08X}{rdevminor:08X}{namesize:08X}{check:08X}",
        ino = 1,
        mode = mode,
        uid = 0,
        gid = 0,
        nlink = 1,
        mtime = 0,
        filesize = data.len(),
        devmajor = 0,
        devminor = 0,
        rdevmajor = 0,
        rdevminor = 0,
        namesize = namesize,
        check = 0,
    )?;
    buf.extend_from_slice(name.as_bytes());
    buf.push(0);
    pad4(buf);
    buf.extend_from_slice(data);
    pad4(buf);
    Ok(())
}

fn pad4(buf: &mut Vec<u8>) {
    if buf.len() % 4 != 0 {
        buf.resize(buf.len() + (4 - buf.len() % 4), 0);
    }
}

#[cfg(test)]
mod tests;
